go/vt/automation/scheduler.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2015, Google Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+/*
+Package automation contains code to execute high-level cluster operations
+(e.g. resharding) as a series of low-level operations
+(e.g. vtctl, shell commands, ...).
+*/
+package automation
+
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/golang/glog"
+	pb "github.com/youtube/vitess/go/vt/proto/automation"
+	"golang.org/x/net/context"
+)
+
+type schedulerState int32
+
+const (
+	stateNotRunning schedulerState = iota
+	stateRunning
+	stateShuttingDown
+	stateShutdown
+)
+
+// Scheduler executes automation tasks and maintains the execution state.
+type Scheduler struct {
+	idGenerator IDGenerator
+
+	mu sync.Mutex
+	// Guarded by "mu".
+	registeredClusterOperations map[string]bool
+	// Guarded by "mu".
+	toBeScheduledClusterOperations chan *ClusterOperationInstance
+	// Guarded by "mu".
+	state schedulerState
+
+	pendingOpsWg *sync.WaitGroup
+
+	muOpList sync.Mutex
+	// Guarded by "muOpList".
+	activeClusterOperations map[string]*ClusterOperationInstance
+	// Guarded by "muOpList".
+	finishedClusterOperations map[string]*ClusterOperationInstance
+}
+
+// NewScheduler creates a new instance.
+func NewScheduler() (*Scheduler, error) {
+	defaultClusterOperations := map[string]bool{
+		"ReshardingTask": true,
+		// Testing only cluster operations.
+		"TestingEchoTask":     true,
+		"TestingEmitEchoTask": true,
+	}
+
+	s := &Scheduler{
+		registeredClusterOperations:    defaultClusterOperations,
+		idGenerator:                    IDGenerator{},
+		toBeScheduledClusterOperations: make(chan *ClusterOperationInstance, 10),
+		state:                     stateNotRunning,
+		pendingOpsWg:              &sync.WaitGroup{},
+		activeClusterOperations:   make(map[string]*ClusterOperationInstance),
+		finishedClusterOperations: make(map[string]*ClusterOperationInstance),
+	}
+
+	return s, nil
+}
+
+// Run processes queued cluster operations.
+func (s *Scheduler) Run() {
+	s.mu.Lock()
+	s.state = stateRunning
+	s.mu.Unlock()
+
+	s.pendingOpsWg.Add(1)
```

Comment:
In the future, there may be more than one operation.

Given the "may" in this situation, I tried to change the code to your suggestion, it would make the overall code more complicated because of the following corner case:

Let's assume this execution:
s := NewScheduler()
s.ShutdownAndWait()

ShutdownAndWait() shouldn't block wait for the "done" channel to be closed now because the Scheduler is not in a "Running" or "ShuttingDown" state. But I would say these extra checks make the code more complicated.

Instead, I've moved the code pendingOpsWg.Add(1) and the launch of the go function into an extra method:

startProcessRequestsLoop()

Now it should be more clear that there may be more than one process.